vault/resource_gcp_auth_backend_role.go

```diff
@@ -0,0 +1,218 @@
+package vault
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/hashicorp/vault/api"
+)
+
+func gcpAuthBackendRoleResource() *schema.Resource {
+	return &schema.Resource{
+		SchemaVersion: 1,
+
+		Create: gcpAuthResourceWrite,
+		Update: gcpAuthResourceUpdate,
+		Read:   gcpAuthResourceRead,
+		Delete: gcpAuthResourceDelete,
+
+		Schema: map[string]*schema.Schema{
+			"role": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"project_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"ttl": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"max_ttl": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"period": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"policies": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				Computed: true,
+			},
+			"bound_service_accounts": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				Computed: true,
+			},
+			"backend": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "gcp",
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func gcpRoleResourcePath(backend, role string) string {
+	return "auth/" + strings.Trim(backend, "/") + "/role/" + strings.Trim(role, "/")
+}
+
+func gcpAuthResourceWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	role := d.Get("role").(string)
+
+	path := gcpRoleResourcePath(backend, role)
+
+	data := map[string]interface{}{}
+
+	if v, ok := d.GetOk("type"); ok {
+		data["type"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("project_id"); ok {
+		data["project_id"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("ttl"); ok {
+		data["ttl"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("max_ttl"); ok {
+		data["max_ttl"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("period"); ok {
+		data["period"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("policies"); ok {
+		data["policies"] = v.(*schema.Set).List()
+	}
+
+	if v, ok := d.GetOk("bound_service_accounts"); ok {
+		data["bound_service_accounts"] = v.(*schema.Set).List()
+	}
+
+	log.Printf("[DEBUG] Writing role %q to GCP auth backend", path)
+	_, err := client.Logical().Write(path, data)
+	d.SetId(path)
```

Comment:
Would it be possible to move this up a line so it's not between the err and the err check?